File: paleo_workbench/ui/pages/preview_provider.py
from __future__ import annotations

import copy
from pathlib import Path

try:
    import segyio
except ImportError:  # pragma: no cover
    segyio = None

from paleo_workbench.project.models import ExportArtifact, ResourceItem
from paleo_workbench.resources.preview_parsers import (
    AUDIO_FORMATS,
    EXCEL_FORMATS,
    GEOTIFF_FORMATS,
    HTML_FORMATS,
    IMAGE_FORMATS,
    JSON_ARRAY_COLLAPSE_THRESHOLD,
    JSON_FORMATS,
    LAS_FORMATS,
    MAX_JSON_PARSE_BYTES,
    MAX_TABLE_COLUMNS,
    MAX_TABLE_ROWS,
    MAX_TEXT_PREVIEW_BYTES,
    MARKDOWN_FORMATS,
    PDF_FORMATS,
    SEGY_FORMATS,
    TABLE_FORMATS,
    TEXT_FORMATS,
    PreviewMode,
    PreviewResult,
)
from paleo_workbench.resources.preview_parsers.registry import default_registry
from paleo_workbench.ui.pages.preview_settings import PreviewSettings

__all__ = [
    "AUDIO_FORMATS",
    "EXCEL_FORMATS",
    "GEOTIFF_FORMATS",
    "HTML_FORMATS",
    "IMAGE_FORMATS",
    "JSON_ARRAY_COLLAPSE_THRESHOLD",
    "JSON_FORMATS",
    "LAS_FORMATS",
    "MAX_JSON_PARSE_BYTES",
    "MAX_TABLE_COLUMNS",
    "MAX_TABLE_ROWS",
    "MAX_TEXT_PREVIEW_BYTES",
    "MARKDOWN_FORMATS",
    "PDF_FORMATS",
    "SEGY_FORMATS",
    "TABLE_FORMATS",
    "TEXT_FORMATS",
    "PreviewMode",
    "PreviewProvider",
    "PreviewResult",
]


class PreviewProvider:
    def __init__(self, settings: PreviewSettings | None = None) -> None:
        self.settings = settings or PreviewSettings.defaults()

    def with_settings(self, settings: PreviewSettings) -> "PreviewProvider":
        """Return a request-local provider snapshot without copying its engine."""
        configured = copy.copy(self)
        configured.settings = settings
        return configured

    def preview(self, asset: ResourceItem | ExportArtifact | None) -> PreviewResult:
        if asset is None:
            return PreviewResult(
                mode="empty",
                title="请选择数据项",
                message="从列表中选择一个数据、成果或文件",
            )

        # Pure build — safe for worker threads. LRU cache is owned by
        # PreviewRequestController on the UI thread.
        return self._build_preview(asset)

    def preview_summary(
        self,
        asset: ResourceItem | ExportArtifact | None,
    ) -> PreviewResult:
        """Build the lightweight reader payload for an asset."""
        return self.preview(asset)

    def preview_visualization(
        self,
        asset: ResourceItem | ExportArtifact | None,
    ) -> PreviewResult:
        """Return a stable result when no professional backend is available."""
        if asset is None:
            return self.preview(None)
        if isinstance(asset, ResourceItem):
            title = asset.name
            path = asset.path
            fmt = asset.format
            status = asset.status
            type_label = asset.type
        else:
            title = Path(asset.output_path).name or asset.output_path
            path = asset.output_path
            fmt = asset.format
            status = "generated"
            type_label = "成果"
        return PreviewResult(
            mode="message",
            title=title,
            path=path,
            format=fmt,
            status=status,
            type_label=type_label,
            message="此数据不支持可视化预览",
        )

    def _build_preview(self, asset: ResourceItem | ExportArtifact) -> PreviewResult:
        return default_registry().build_preview(asset, self.settings)

File: libs/ui_data_core/src/preview_provider.cpp
#include "pwb/ui_data_core/preview_provider.hpp"

#include "pwb/ui_data_core/json_util.hpp"

#include <stdexcept>
#include <utility>

namespace pwb::ui_data_core {

PreviewResult empty_preview_result() {
    PreviewResult result;
    result.mode = preview_mode::kEmpty;
    result.title = "请选择数据项";
    result.message = "从列表中选择一个数据、成果或文件";
    return result;
}

PreviewProvider::PreviewProvider(BuildFn builder, PreviewSettings settings)
    : builder_(std::move(builder)), settings_(std::move(settings)) {}

PreviewResult PreviewProvider::preview(const AssetObjectData* asset) const {
    if (asset == nullptr) {
        return empty_preview_result();
    }
    // _build_preview → default_registry().build_preview(asset, settings).
    return builder_(*asset, settings_);
}

PreviewResult PreviewProvider::preview_summary(
    const AssetObjectData* asset) const {
    return preview(asset);
}

PreviewResult PreviewProvider::preview_visualization(
    const AssetObjectData* asset) const {
    if (asset == nullptr) {
        return preview(nullptr);
    }
    PreviewResult result;
    result.mode = preview_mode::kMessage;
    result.message = "此数据不支持可视化预览";
    if (const auto* resource = std::get_if<ResourceItem>(asset)) {
        result.title = resource->name;
        result.path = resource->path;
        result.format = resource->format;
        result.status = resource->status;
        result.type_label = resource->type;
        return result;
    }
    if (const auto* artifact = std::get_if<ExportArtifact>(asset)) {
        // Path(output_path).name or output_path — a bare-name fallback.
        const std::string name = python_path_name(artifact->output_path);
        result.title = name.empty() ? artifact->output_path : name;
        result.path = artifact->output_path;
        result.format = artifact->format;
        result.status = "generated";
        result.type_label = "成果";
        return result;
    }
    // The Python type contract admits only ResourceItem | ExportArtifact;
    // any other variant mirrors the AttributeError branch.
    throw std::invalid_argument(
        "preview_visualization: unsupported asset variant");
}

}  // namespace pwb::ui_data_core
